### PR TITLE
Implement etcd snapshot restore skeleton

### LINKED ISSUE #500: [etcdsnapshot&restore] Implement ETCDSnapshotRestore controller
For implementation details please refer to the proposal and POC code: https://github.com/furkatgofurov7/turtles-etcd-restore/blob/main/internal/controller/etcdsnapshotrestore_controller.go

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #500 

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits into logical changes
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

### CODE REVIEW

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go [hunk 1]
```diff
@@ -19,65 +19,381 @@ package controllers
...
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 
 	snapshotrestorev1 "github.com/rancher/turtles/exp/etcdrestore/api/v1alpha1"
 )
 
-// EtcdSnapshotRestoreReconciler reconciles a EtcdSnapshotRestore object.
-type EtcdSnapshotRestoreReconciler struct {
+// InitMachine is a filter matching on init machine of the ETCD snapshot
+func InitMachine(etcdMachineSnapshot *snapshotrestorev1.ETCDMachineSnapshot) collections.Func {
```
Comment (alexander-demicev) [inline exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:43]: can we make it private?

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go [hunk 2]
```diff
@@ -19,65 +19,381 @@ package controllers
...
+		return machine.Status.Phase == string(clusterv1.MachinePhaseRunning)
+	}
+
+	if scope.machines.Filter(running).Len() != scope.machines.Len() {
+		log.Info("Not all machines are running yet, requeuing")
+		return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+	}
+
+	if scope.machines.Filter(InitMachine(scope.etcdMachineSnapshot)).Len() != 1 {
+		return ctrl.Result{}, fmt.Errorf(
```
Comment (alexander-demicev) [inline exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:163]: does it make sense to mark restore as failed if machine doesn't exist?

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go [hunk 3]
```diff
@@ -19,65 +19,381 @@ package controllers
...
+		machines:            machines.Filter(collections.ControlPlaneMachines(cluster.Name)),
+		etcdMachineSnapshot: etcdMachineSnapshot,
+	}, nil
+}
+
+func (r *ETCDSnapshotRestoreReconciler) stopRKE2OnAllMachines(ctx context.Context, scope *scope, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	results := []Output{}
+	for _, machine := range scope.machines {
```
Comment (alexander-demicev) [inline exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:258]: based on recent discussions, we can stop RKE2 service only on control planes

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go [hunk 4]
```diff
@@ -19,65 +19,381 @@ package controllers
...
 
-func (r *EtcdSnapshotRestoreReconciler) reconcileNormal(
-	_ context.Context, _ *snapshotrestorev1.EtcdSnapshotRestore,
-) (_ ctrl.Result, err error) {
+func (r *ETCDSnapshotRestoreReconciler) startRKE2OnAllMachines(ctx context.Context, scope *scope, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	initMachine := scope.machines.Filter(InitMachine(scope.etcdMachineSnapshot)).UnsortedList()[0]
+
+	// TODO: other registration methods
```
Comment (alexander-demicev) [inline exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:323]: we should probably create an issue

File: exp/etcdrestore/api/v1alpha1/etcdsnapshotrestore_types.go [hunk 5]
```diff
@@ -44,41 +45,53 @@ const (
...
+	// +required
+	ClusterName string `json:"clusterName"`
+
+	// +required
+	ETCDMachineSnapshotName string `json:"etcdMachineSnapshotName"`
+
+	// TTLSecondsAfterFinished int `json:"ttlSecondsAfterFinished"`
+
+	// // +required
+	// ConfigRef corev1.LocalObjectReference `json:"configRef"`
```
Comment (alexander-demicev) [inline exp/etcdrestore/api/v1alpha1/etcdsnapshotrestore_types.go:62]: We need an interface to support K3S in future

**Review by alexander-demicev — APPROVED** [review]
Assuming this is WIP no comments are blocking merging the PR, thanks a lot for taking care of it @Danil-Grigorev

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go (hunk 2 above)
Comment (Danil-Grigorev) [reply exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:163]: It does, however the machine may go away after the restore. This needs to return early if the restore was completed

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go (hunk 3 above)
Comment (Danil-Grigorev) [reply exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:258]: I think currently the set of machines already limited to CP. Only agent installation is performed on worker nodes.

File: exp/etcdrestore/api/v1alpha1/etcdsnapshotrestore_types.go [hunk 6]
```diff
@@ -44,41 +45,53 @@ const (
...
+//
+// ETCDSnapshotRestoreSpec defines the desired state of EtcdSnapshotRestore.
+type ETCDSnapshotRestoreSpec struct {
+	// +required
+	ClusterName string `json:"clusterName"`
+
+	// +required
+	ETCDMachineSnapshotName string `json:"etcdMachineSnapshotName"`
+
+	// TTLSecondsAfterFinished int `json:"ttlSecondsAfterFinished"`
```
Comment (furkatgofurov7) [inline exp/etcdrestore/api/v1alpha1/etcdsnapshotrestore_types.go:59]: why are they commented out?

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go [hunk 7]
```diff
@@ -19,65 +19,381 @@ package controllers
...
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseStarted:
+		etcdSnapshotRestore.Status.Phase = snapshotrestorev1.ETCDSnapshotRestorePhaseShutdown
+
+		return ctrl.Result{}, nil
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseShutdown:
+		// Stop RKE2 on all the machines.
+		return r.stopRKE2OnAllMachines(ctx, scope, etcdSnapshotRestore)
+	case snapshotrestorev1.ETCDSnapshotRestorePhaseRunning:
+		// Restore the etcd snapshot on the init machine.
+		return r.restoreSnaphotOnInitMachine(ctx, scope, etcdSnapshotRestore)
```
Comment (furkatgofurov7) [inline exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:202]:
```suggestion
		return r.restoreSnapshotOnInitMachine(ctx, scope, etcdSnapshotRestore)
```

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go [hunk 8]
```diff
@@ -19,65 +19,381 @@ package controllers
...
+	}
+
+	log.Info("All machines are ready to proceed to the next phase, setting phase to running")
+
+	etcdSnapshotRestore.Status.Phase = snapshotrestorev1.ETCDSnapshotRestorePhaseRunning
+
+	return ctrl.Result{}, nil
+}
+
+func (r *ETCDSnapshotRestoreReconciler) restoreSnaphotOnInitMachine(ctx context.Context, scope *scope, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (ctrl.Result, error) {
```
Comment (furkatgofurov7) [inline exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:289]:
```suggestion
func (r *ETCDSnapshotRestoreReconciler) restoreSnapshotOnInitMachine(ctx context.Context, scope *scope, etcdSnapshotRestore *snapshotrestorev1.ETCDSnapshotRestore) (ctrl.Result, error) {
```

**Review by furkatgofurov7 — APPROVED** [review]
Thanks for working on this @Danil-Grigorev

File: exp/etcdrestore/api/v1alpha1/etcdsnapshotrestore_types.go (hunk 6 above)
Comment (Danil-Grigorev) [reply exp/etcdrestore/api/v1alpha1/etcdsnapshotrestore_types.go:59]: We should address this as a part of another issue #764

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go (hunk 7 above)
Comment (Danil-Grigorev) [reply exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:202]: I’m addressing this as a part of #769

File: exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go (hunk 1 above)
Comment (Danil-Grigorev) [reply exp/etcdrestore/controllers/etcdsnapshotrestore_controller.go:43]: Addressed as a part of #769

**Pull request merged**